docs(react-infobutton): Adding component's spec (#25251)
microsoft/fluentui: packages/react-components/react-infobutton/Spec.md

# @fluentui/react-infobutton Spec

Convergence epic issue: [#25062](https://github.com/microsoft/fluentui/issues/25062)

> ## Background

Reviewer: Can you please add some visuals to show how component looks?

> An InfoButton provides a way for users to get more information about a particular UI element. It is a button that displays an icon and when clicked a Popover displays the additional information.

Reviewer: Do I correctly understand that this component is basically following?

```jsx
<Popover>
  <PopoverTrigger>
    <Button icon='X' />
  </PopoverTrigger>
  <PopoverSurface>Some content</PopoverSurface>
</Popover>
```

Author: After an update, we are not using the v9 button, but the native button. Other than that, you are correct. It's a pretty simple component, but will bring an opinionated approach to it required for Field.

Reviewer 2: This paragraph should be updated to make it obvious that InfoButton can contain focusable content as well (Link/button etc)

Because the Popover isn't always visible, it should not contain information that people must know in order to complete the field.

## Prior Art

_Include background research done for this component_

> #### Open UI

Reviewer: This section is not related to OpenUI, it should be under "Prior Art", correct?

| TextField with example on how to add an info button | [v8](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield)          |       |
| ContextualHelp                                      | [React Spectrum](https://react-spectrum.adobe.com/react-spectrum/ContextualHelp.html) |       |

#### Comparison with v0 and v8

- v0 does not have an InfoButton component.
- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see Open UI section above for more details.

## Sample Code

```jsx
<InfoButton content="This is some additional information." />
```

## Variants

Visual (only the button has a variant)

- `transparent` (default)
- `subtle`

## API

#### Props

```ts
import type { Button } from '@fluentui/react-button';
import type { PopoverProps, PopoverSurface } from '@fluentui/react-popover';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type InfoButtonSlots = {
  root: NonNullable<Slot<PopoverProps>>;

  /**
   * The button that triggers the Popover.
   */

>   trigger: NonNullable<Slot<typeof Button>>;

Reviewer: Let's avoid calling this `trigger` to avoid confusion with the triggers pattern.

  /**
   * The content to be displayed in the Popover.
   */
  content: NonNullable<Slot<typeof PopoverSurface>>;
};

/**
 * InfoButton Props
 */

> export type InfoButtonProps = Omit<ComponentProps<Partial<InfoButtonSlots>>, 'children'>;

Reviewer: Is there is no `children`, how can I customize `button`'s content? `<InfoButton trigger={{ children: 'X' }} />`?

Author: Correct, it would be through the slot. This omit is only removing the children for the root slot (Popover). Another option is to make children be the children of content, but I decided against that due to content being a slot itself.

/**
 * State used in rendering InfoButton
 */
export type InfoButtonState = ComponentState<InfoButtonSlots> & {

>   popoverOpen: boolean;

Reviewer: Will be this exposed in `props`? If no, how can I control `Popover`'s state?

Author: That can be done directly on the props like this:
`<InfoButton open={controlledOpen} />`

};
```

## Structure

_**Public**_

```jsx
<InfoButton content="This is some additional information." />

> ```

Reviewer: Is it possible to attach a `Tooltip` to it? I.e.

```jsx
<Tootlip>
  <InfoButton content="This is some additional information." />
</Tooltip>
```

Author: I'm not sure why someone would do that in this case, but it should be possible.

| TextField with example on how to add an info button | [v8](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield)          | This is just an example and on how to implement it.                                                      |
| ContextualHelp                                      | [React Spectrum](https://react-spectrum.adobe.com/react-spectrum/ContextualHelp.html) | It's a full component and uses a Popover instead of tooltip.                                             |

#### Comparison with v0 and v8

- v0 does not have an InfoButton component.
- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see Open UI section above for more details.

## Sample Code

```jsx
<InfoButton content="This is some additional information." />
```

## Variants

Visual (only the button has a variant)

- `transparent` (default)

![transparent](./etc/images/transparent-infobutton.png)

## API

#### Props

```ts
export type InfoButtonSlots = {
  root: NonNullable<Slot<'button'>>;

  /**
   * The PopoverSurface to be displayed when the button is pressed.
   */

>   popoverSurface: NonNullable<Slot<typeof PopoverSurface>>;

Reviewer 3: Can we call this `content` so it is similar to Tooltip?

};

/**
 * InfoButton Props
 */

> export type InfoButtonProps = ComponentProps<Partial<InfoButtonSlots>> & Omit<PopoverProps, 'children'>;

Reviewer 3: This should pick specific popover props, rather than taking all of them (minus children). The code in useInfoButton needs to know about each Popover prop that it's forwarding. If a new prop is added to Popover in the future, that would implicitly add it to InfoButton, but it wouldn't be functional. So it'd be safer to explicitly list the Popover props that are supported by InfoButton.

Reviewer 3: That said, I don't think we typically do prop forwarding like this. It'd be a little less ergonomic, but I think having a `popover` slot would be more in line with the v9 API design. That would also avoid having to manually pick and forward props to the popover.

```

## Structure

_**Public**_

```jsx
<InfoButton
  popoverSurface={
    <>
      Popover above-start lorem ipsum dolor sit amet consectetur.{' '}
      <Link href="https://react.fluentui.dev">Learn more</Link>
    </>
  }
/>
```

_**Internal**_

```jsx
return (

>   <Popover {...(state as Omit<PopoverProps, 'children'>)}>

Reviewer 3: I suppose this is more of a code review comment, but we definitely don't want to blindly spread state like this. You'd need to write out each of the props being forwarded. (Or make `popover` a slot like in my other comment)

- _Component States_
  - Popover open: The Popover is open and content visible.
- _Interaction_
  - _Keyboard_
    - `Enter` or `Space` key: Opens the Popover.
      - Focusable items in Popover: Item should trap focus within the Popover.
      - No focusable items in Popover: Focus should stay on the button.
    - `Escape` key: Closes the Popover.
  - _Cursor_
    - `Click`: Opens the Popover.
    - `Click` outside of Popover: Closes the Popover.
  - _Touch_
    - `Tap`: Opens the Popover.
    - `Tap` outside of Popover: Closes the Popover.
  - _Screen readers_
    - When screen reader is on the button, it should announce that it is a button, that it can be used to open a Popover, and read out the aria-label of the button.
    - When screen reader is on the Popover, it should announce that it is a dialog, and read out the content of the Popover.

## Accessibility

> - `role="tooltip"` is used on the PopoverSurface.

Reviewer 3: Do we then also need to use `aria-describedby` or `aria-labeledby` on the button itself?

Author: This isn't set yet, once we iron it out, I'll update the spec.

# @fluentui/react-infobutton Spec

Convergence epic issue: [#25062](https://github.com/microsoft/fluentui/issues/25062)

## Background

An InfoButton provides a way for users to get more information about a particular UI element. It is a button that displays an icon and when clicked a Popover displays the additional information.

Because the Popover isn't always visible, it should not contain information that people must know in order to complete the field.

### Anatomy

> ![Anatomy](./etc/images/anatomy.png)

Reviewer 4: This image shows the popover in the "above-start" position. Will `InfoButton` support positioning of the popover?

Author: Yes, that will be done through the root as it takes the popover props as well. The updated types now have the Popover props.

| Input and guidance to add info button               | [Salesforce](https://www.lightningdesignsystem.com/components/input/)                 | This component lives within Input and shows how to build an info button instead.                         |
| TextField with example on how to add an info button | [v8](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield)          | This is just an example and on how to implement it.                                                      |
| ContextualHelp                                      | [React Spectrum](https://react-spectrum.adobe.com/react-spectrum/ContextualHelp.html) | It's a full component and uses a Popover instead of tooltip.                                             |

#### Comparison with v0 and v8

- v0 does not have an InfoButton component.
- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see Open UI section above for more details.

## Sample Code

```jsx

> <InfoButton content="This is some additional information." />

Reviewer 4: One of the images shows a link in the popover. Would be nice to have a code sample showing you can use JSX in the `content` prop to make it clear we `InfoButton` supports that.

## API

#### Props

```ts
export type InfoButtonSlots = {
  root: NonNullable<Slot<PopoverProps>>;

  /**
   * The button that triggers the Popover.
   */
  button: NonNullable<Slot<'button'>>;

  /**
   * The content to be displayed in the Popover.
   */
  content: NonNullable<Slot<typeof PopoverSurface>>;
};

/**
 * InfoButton Props
 */
export type InfoButtonProps = Omit<ComponentProps<Partial<InfoButtonSlots>>, 'children'>;

/**
 * State used in rendering InfoButton
 */
export type InfoButtonState = ComponentState<InfoButtonSlots> & {
  popoverOpen: boolean;
};
```

## Structure

_**Public**_

```jsx
<InfoButton content="This is some additional information." />
```

_**Internal**_

```jsx
return (
  <slots.root {...(slotProps.root as PopoverProps)}>
    <PopoverTrigger>
      <slots.button {...slotProps.button} />
    </PopoverTrigger>
    <slots.content {...slotProps.content} />
  </slots.root>
);
```

_**DOM**_

```html
<button type="button" class="fui-Button fui-InfoButton__button">
  <!-- icon -->
</button>

<!-- on document.body -->
<div role="tooltip" class="fui-PopoverSurface fui-InfoButton__content">This is some additional information.</div>
```

## Migration

> There's no migration guide as `v0` and `v8` do not have an InfoButton component.

Reviewer 5: Hello :)
there was no similar component in `v0` or in 'v8' but one version exists in the TMP repository(in MS Teams code)

Implementation accessibility details are following:
- icon is focusable
- icon has role=img
- with tooltip is icon connected via "aria-describedby" approach
- icon has label "More information"
- tooltip is displayed when icon receive focus
https://domoreexp.visualstudio.com/Teamspace/_git/teams-modular-packages/pullrequest/477885

This approach was approved by accessibility PM for MS Teams(Brett Humphrey).

I believe we should be aligned together with v9 and MS Teams component.

# @fluentui/react-infobutton Spec

Convergence epic issue: [#25062](https://github.com/microsoft/fluentui/issues/25062)

## Background

> An InfoButton provides a way for users to get more information about a particular UI element. It is a button that displays an icon and when clicked a Popover displays the additional information. InfoButton may contain focusable items inside the Popover.

Reviewer 6: _nit:_ wording

```suggestion
An InfoButton provides a way for users to get more information about a particular UI element. It is a button containing an icon that, when clicked, displays a Popover with the additional information. InfoButton may contain focusable items inside the Popover.
```

| Tooltip with card appearance and general tooltip    | [EverGreen](https://evergreen.segment.com/components/tooltip)                         | This are examples within the tooltip component. This library doesn't use a button, instead just an icon. |
| Input and guidance to add info button               | [Salesforce](https://www.lightningdesignsystem.com/components/input/)                 | This component lives within Input and shows how to build an info button instead.                         |
| TextField with example on how to add an info button | [v8](https://developer.microsoft.com/en-us/fluentui#/controls/web/textfield)          | This is just an example and on how to implement it.                                                      |
| ContextualHelp                                      | [React Spectrum](https://react-spectrum.adobe.com/react-spectrum/ContextualHelp.html) | It's a full component and uses a Popover instead of tooltip.                                             |

#### Comparison with v0 and v8

- v0 does not have an InfoButton component.

> - v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see Open UI section above for more details.

Reviewer 6: ```suggestion
- v8 does not have an InfoButton component, but does have a TextField component that has an example of how to add an info button to a TextField, see "Prior Art" section above for more details.
```

## Sample Code

```jsx
<InfoButton content="This is some additional information." />
```

## Variants

Visual (only the button has a variant)

- `transparent` (default)

> ![transparent](./etc/images/transparent-infobutton.png)

Reviewer 3: InfoButtons are _always_ transparent right? So it's not technically a variant 🙂 it's just how InfoButtons look.

Author: It hasn't been finalized, but we might add subtle. I'll remove it for now!

## API

#### Props

```ts
export type InfoButtonSlots = {
  root: NonNullable<Slot<'button'>>;

  /**
   * The Popover element that wraps the content and root. Use this slot to pass props to the Popover.
   */
  popover: NonNullable<Slot<PopoverProps>>;

  /**
   * The content to be displayed in the PopoverSurface when the button is pressed.
   */
  content: NonNullable<Slot<typeof PopoverSurface>>;
};

/**
 * InfoButton Props
 */
export type InfoButtonProps = ComponentProps<Partial<InfoButtonSlots>>;

/**
 * State used in rendering InfoButton
 */
export type InfoButtonState = ComponentState<InfoButtonSlots>;
```

## Structure

_**Public**_

```jsx
<InfoButton
  content={
    <>

>       Popover above-start lorem ipsum dolor sit amet consectetur.{' '}

Reviewer 2: ```suggestion
      Popover above-start lorem ipsum dolor sit amet consectetur.
```

>       <Link href="https://react.fluentui.dev">Learn more</Link>

Reviewer 2: Is focus going to be handled to allow actionable items in the popover? I don't recall that being addressed in the description above.

Reviewer 2: I see it mentioned below, but it might be good to call out in the initial description above (can contain focusable content)

    </>
  }
/>
```

_**Internal**_

```jsx
return (
  <slots.popover {...(slotProps.popover as PopoverProps)}>
    <PopoverTrigger>
      <slots.root {...slotProps.root} />
    </PopoverTrigger>
    <slots.content {...slotProps.content} />
  </slots.popover>
);
```

_**DOM**_

```html
<button type="button" class="fui-Button fui-InfoButton__button">
  <!-- icon -->
</button>

<!-- on document.body -->
<div role="tooltip" class="fui-PopoverSurface fui-InfoButton__content">
  Popover above-start lorem ipsum dolor sit amet consectetur.
  <a href="https://react.fluentui.dev">Learn more</a>
</div>
```

## Migration

There's no migration guide as `v0` and `v8` do not have an InfoButton component.

## Behaviors

> > Note: The behavior will change when used in conjunction with the `Field` component. It is mentioned that when a form has many infobuttons, it may be better to not make a full tab stop in the button. The idea is to have each field have a "shortcut" that will let you focus on the infobutton. This will be implemented by Field if needed.

Reviewer 6: ```suggestion
> Note: The behavior will change when used in conjunction with the `Field` component. It is mentioned that when a form has many InfoButtons, it may be better to not make a full tab stop in the button. The idea is to have each field have a "shortcut" that will let you focus on the infobutton. This will be implemented by Field if needed.
```

- _Component States_
  - Popover open: The Popover is open and content visible.
- _Interaction_
  - _Keyboard_
    - `Enter` or `Space` key: Opens the Popover.
      - Focusable items in Popover: Item should trap focus within the Popover.
      - No focusable items in Popover: Focus should stay on the button.
    - `Escape` key: Closes the Popover.
  - _Cursor_
    - `Click`: Opens the Popover.
    - `Click` outside of Popover: Closes the Popover.
  - _Touch_
    - `Tap`: Opens the Popover.
    - `Tap` outside of Popover: Closes the Popover.
  - _Screen readers_
    - When screen reader is on the button, it should announce that it is a button, that it can be used to open a Popover, and read out the aria-label of the button.
    - When screen reader is on the Popover, it should announce that it is a dialog, and read out the content of the Popover.

## Accessibility

- `role="tooltip"` is used on the PopoverSurface.
- Tab order
  - When tabbing through the page, the button should be a tab stop.
  - When focused on the button, pressing `Enter` or `Space` should open the Popover.

>     - Focus is not trapped within the Popover, it should stay in the button.

Reviewer 2: ```suggestion
    - If the popover does not contain any focusable items, focus should stay on the button.
```
